allennlp/commands/train.py

```diff
@@ -51,6 +51,9 @@ def add_subparser(parser: argparse._SubParsersAction) -> argparse.ArgumentParser
                            type=str,
                            required=True,
                            help='directory in which to save the model and its logs')
+    subparser.add_argument('-o', '--overrides',
```

Comment:
I think this is great for `evaluate` and `predict`, but I'm conflicted about allowing this for `train`.  I kind of think this shouldn't even be an option, because it encourages you to do the wrong thing.

Reply:
I understand your point, but I'd like to leave it in:

* It's available for the other options.
* The code is already written.
* We may come up with a use case.

Reply:
Using this parameter during training actually breaks things.  The way that we archive models is by copying the config file that was used for training, and using it when loading the model.  The overrides that you passed to `train` will not be present, and thus your model will be broken.

Incidentally, this also means that using environment variables _also_ breaks things.  We really should just have one way to specify parameters to `train`, not ways to override a file that you can just directly edit.

Reply:
This doesn't seem to be a problem.  We save from the model from what we have in memory rather than copying the original file.

https://github.com/allenai/allennlp/blob/master/allennlp/commands/train.py#L113

Here's a snippet to verify this behavior.  It also substitutes variables with environment variables in the file it saves.

```
import json
import os
from allennlp.common import Params
from copy import deepcopy

filename = 'tests/fixtures/bidaf/experiment.json'
params = Params.from_file(filename, '{ foo: $PATH }')
serialization_params = deepcopy(params).as_dict(quiet=True)
with open("model_params_test.json", "w") as param_file:
    json.dump(serialization_params, param_file, indent=4)
```

Reply:
Well, that's better than I thought.  I'm still not a fan of having this as an option for training; having the same functionality in more than one place is confusing and error prone.